book_src/web_stuff/web_gl_with_bare_wasm.md

```diff
@@ -0,0 +1,683 @@
+
+> cargo install devserver
+
+If you already have your own favorite way to spin up a local server that can serve static files, that's fine too.
+
+## Separate Folder
+
+This will have a new non-standard requirements,
+so I'm going to put it in a `web_crate/` directory.
+
+First it needs its own `Cargo.toml` file:
+```toml
+[package]
+name = "triangle-from-scratch-web-crate"
+version = "0.1.0"
+authors = ["Lokathor <zefria@gmail.com>"]
+edition = "2018"
+license = "Zlib OR Apache-2.0 OR MIT"
+```
+
+Now also, to make a wasm library like we need we have to tell Rust that the [crate-type](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#the-crate-type-field)
+will be `cdylib`:
+```toml
+[lib]
+crate-type = ["cdylib"]
+```
+
+Personally I also like to turn on [link-time optimization](https://doc.rust-lang.org/cargo/reference/profiles.html#lto) with release builds,
+not because it's required,
+but just because I'm willing to spend some extra compile time to get a performance edge.
+The winner here is "thin",
+which provides almost all the benefit for a minimal amount of additional time and memory taken to compile.
+
+```toml
+[profile.release]
+lto = "thin"
+```
+
+Now we're set.
+
+## The Wasm Library
+
+As you can sorta already see, our "program" isn't actually going to be built as an executable.
+Instead, it's going to be built as a C-compatible library that the JavaScript of the webpage will load and use.
+This means that instead of writing a `main.rs` with an optional `lib.rs`,
+we put 100% of the code into `lib.rs` right from the start.
+
+```rust
+// lib.rs
+
+#[no_mangle]
+pub extern "C" fn start() {
+  // nothing yet!
+}
+```
+
+Note the use of the [no_mangle](https://doc.rust-lang.org/reference/abi.html#the-no_mangle-attribute) attribute.
+This totally disables the usual name mangling that Rust does.
+It allows for the function to be called by external code that doesn't know Rust's special naming scheme, which is good,
+but there can only be a single function with a given name anywhere.
+In other words, if some other function named `start` with no mangling existed *anywhere* in our project,
+or in any of our dependencies, then we'd get a compilation error.
+That's why name mangling is on by default.
+
+Also note that we have to declare that our `start` function uses the `extern "C"` ABI,
+this will give us the correct calling convention when communicating between JavaScript and Wasm.
+
+When JavaScript loads up our wasm module, the `start` function will be called.
+This will allow our program to do whatever it wants to do,
+similar to the `main` function in a normal program.
+
+## The Web Page
+
+Okay now we need a webpage for the user to display and have the wasm go.
+
+I'm absolutely not a web development person,
+but I know just enough to throw some HTML together by hand:
+
+```html
+<html>
+<canvas width="800" height="600" id="my_canvas"></canvas>
```

Comment:
The `canvas` should go in the `body`, ideally. Web browsers are fairly generous about what weird and wonderful HTML they'll accept, though, so it wouldn't surprise me if this still worked.

Reply:
Ah, yeah, "it works on my machine", and that's just how it was in the example i followed.

But we can move it into the body.